### DISCUSSION TITLE
Does pub/sub model available for opcua-asyncio?

### DISCUSSION THREAD

**Discussion by SSaiPrasad7** [FreeOpcUa/opcua-asyncio#408] [Q&A]
Hello opcua-asyncio team.
I am beginner. I have doubts regarding opcua publish/subscribe model.
Is pub/sub model implemented for opcua-asyncio ?
If it is implemented,Can you all guide me where to start?

**Comment by oroulet** [comment]
No. If I had time that would be the second thing I would do, after using dataclass in ua structures

**Reply by SSaiPrasad7** [reply]
Great.All the best.Thanks for the immediate response.

**Comment by swamper123** [comment]
A quote from a famous show wit a sponge: 

Seargent Roderick: "Before you learn to drive, you must learn to crawl. Then you learn to walk and then you learn to run. But before you learn to walk, you mest learn to crawl. I want you to crawl!"

What I mean (and I guess oroulet as well) is, that you need to learn some python basics (especially asyncio), OPC UA basics (how stuff is organised), how both things are working together here in the project and maybe then we can talk about the pub/sub model and how it works.

Jumping headfirst into that thing will cause a brain meltdown, perhaps chocolat addiction (trust me, I was there) and suffering.

**Reply by SSaiPrasad7** [reply]
Hello @swamper123. Thanks for the immediate response and quote.I understood what you are saying but I had completed with python and opcua basics.I got some understanding about both things working together.So,that is the reason I am able to ask the question otherwise I cannot(without basics).

**Reply by swamper123** [reply]
If you say so, feel free to take a look on our examples, especially client-subscription.py

**Reply by SSaiPrasad7** [reply]
Thank you.Sure.But I have a doubt regarding user authentication in Python OPC UA also.Can you please share server and client version examples of user authentication(without certificates)?

**Reply by SSaiPrasad7** [reply]
cc @maneetgoyal

**Reply by AndreasHeine** [reply]
its not for opcua-asyncio but its easy to make the transition to asyncio:
https://github.com/FreeOpcUa/python-opcua/issues/1153

**Reply by SSaiPrasad7** [reply]
Hello @AndreasHeine .I saw the issue,its pretty good in that there is a method named `set_user_manager` in user_manager class,but that method is not available in `opcua-asyncio`. I searched all the `.py` files in server directory for similar kinds of methods.I didn't find the method for authentication.If you think transition can be done,please lead me that way.
